sw/device/tests/autogen/plic_all_irqs_test.c

```diff
@@ -0,0 +1,183 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_rv_plic.h"
+#include "sw/device/lib/dif/dif_uart.h"
+#include "sw/device/lib/handler.h"
+#include "sw/device/lib/irq.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/test_framework/test_main.h"
+#include "sw/device/lib/testing/test_framework/test_status.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+static dif_uart_t uart0;
+static dif_rv_plic_t plic;
+
+static const top_earlgrey_plic_target_t kHart = kTopEarlgreyPlicTargetIbex0;
+
+/**
+ * Flag indicating which peripheral is under test.
+ *
+ * Declared volatile because it is referenced in the main program flow as well
+ * as the ISR.
+ */
+static volatile top_earlgrey_plic_peripheral_t peripheral_expected;
+
+/**
+ * Flag indicating which interrupt fired.
+ *
+ * These are declared volatile since they are referenced in the main program
+ * flow as well as the ISR.
+ */
+static volatile dif_uart_irq_t uart_irq_expected;
+static volatile dif_uart_irq_t uart_irq_serviced;
+
+/**
+ * UART irq handler.
+ */
+static void uart_isr(dif_uart_t *uart, dif_rv_plic_irq_id_t plic_irq_id) {
+  // Correlate the interrupt fired at PLIC with UART.
+  dif_uart_irq_t uart_irq =
+      (dif_uart_irq_t)(plic_irq_id - (dif_rv_plic_irq_id_t)
+                                         kTopEarlgreyPlicIrqIdUart0TxWatermark);
+  CHECK(uart_irq == uart_irq_expected,
+        "Incorrect UART irq received at PLIC: exp = %d, obs = %d",
+        uart_irq_expected, uart_irq);
+  uart_irq_serviced = uart_irq;
+
+  // Check if the same interrupt fired at UART as well.
+  bool is_pending;
+  CHECK(dif_uart_irq_is_pending(uart, uart_irq, &is_pending) == kDifUartOk,
+        "dif_uart_irq_is_pending failed");
+  CHECK(is_pending, "UART interrupt fired at PLIC did not fire at UART");
+
+  // TODO: read the whole status register to confirm none other pending.
+
+  // Clear the interrupt at UART.
+  CHECK(dif_uart_irq_acknowledge(uart, uart_irq) == kDifUartOk,
+        "dif_uart_irq_acknowledge failed");
+}
```

Comment:
@timothytrippel, do you think we could autogen this function for all blocks in the testutils area and a common ISR API that can be used in tests directly?

Reply:
Yes absolutely. Created task in #8142 to track this. I will create a separate PR for this.